### ISSUE TITLE
MultiGraph initialization with dict-of-dict-of-dict-of-dict doesn't work properly

### ISSUE THREAD

**Issue by timebarn** [networkx/networkx#4824]
### Current Behavior
The initialization of a multigraph with dictionaries treats the third dictionary, which its keys suppose to be the edges IDs, as the attribute of the edge.
### Expected Behavior
Initializing a multigraph should not accept dictionaries, or it should use reserved keys for edges' IDs.
### Steps to Reproduce
```
g = networkx.MultiGraph({'a': {'b': {0: {'w': 200}, 1: {'w': 201}}}})
print(g.edges)
h = networkx.MultiGraph()
h.add_edges_from([('a', 'b', {"w": 200}), ('a', 'b', {"w": 201})])
print(h.edges)

output

[('a', 'b', 0)]
[('a', 'b', 0), ('a', 'b', 1)]
```
### Environment
Python version: 3.9.5
NetworkX version: 2.5.1

**Comment by dschult** [comment]
See #4823

**Comment by dschult** [comment]
@timebarn, It might help us decide what API to use for this case of dict-of-dict-of-dict-of-dict input if we knew more about your use case.  Where are these data structures coming from? Would it be tricky or limiting if we made you provide an input flag to state whether the input represents a muiltigraph as opposed to a set of edges attributes with the values being dicts.

The short term workaround is:
```python
MG = nx.to_networkx_graph(dodod, create_using=nx.MultiGraph, multigraph_input=True)
```
as the code currently stands, you can't do it with the `nx.MultiGraph(dodod)` syntax.
#4823 would require 
```python
MG = nx.MultiGraph(dodod, multigraph_input=True)
```

**Comment by timebarn** [comment]
I managed to come across this issue with ``` add_edges_from() ``` and to use ``` list``` as an input to this function.
A flag in the function will not be a good idea since one may sometimes add an attribute that can look like two edges. I think that a "flag" option should be for every edge. For example:
```
nx.MultiGraph({'a':{'b':{'multigraph_flag':True,0:{"attr":"foo},1:{"attr":"bar"}}},'c':{'b':{'multigraph_flag':False (or not specifying) ,0:{"attr":"foo},1:{"attr":"bar"}})
```
That means that from 'a' to 'b' you have 2 edges, and from 'c' to 'b', you have only one edge that this is its attributes.
Thanks for the quick response.

**Comment by MridulS** [comment]
@timebarn I'm not sure if the example provided would be a valid data structure for a networkx `MultiGraph`. An edge can't have two values for the same attribute in the way provided in the example. A way of achieving something like this would be
``` python
In [56]: G = nx.MultiDiGraph()

In [57]: G.add_edge('a', 'b', attr='foo')
Out[57]: 0

In [58]: G.add_edge('a', 'b', attr='bar')
Out[58]: 1

In [59]: G.add_edge('c', 'd', attr={0: 'foo', 1: 'bar'})
Out[59]: 0

In [60]: nx.to_dict_of_dicts(G)
Out[60]:
{'a': {'b': {0: {'attr': 'foo'}, 1: {'attr': 'bar'}}},
 'b': {},
 'c': {'d': {0: {'attr': {0: 'foo', 1: 'bar'}}}},
 'd': {}}
```
The third layer of the dictionary will always be the edge id not the attributes in a `Multi(Di)Graph`.
``` python
{'first_node': 
        {'second_node':
               {'edge_id':
                     {'attribute_1': any object
                      'attribute_2': any object }}}}
```

Let me know if I am misunderstanding this :)

**Comment by timebarn** [comment]
@MridulS 

> The third layer of the dictionary will always be the edge id, not the attributes in a Multi(Di)Graph.

This is exactly what I'm speaking about; when initializing, you can't know if the third dict refers to many edges and the keys are IDs or it refers to just one edge and the keys are attributes. Please look at my issue up above.

**Comment by MridulS** [comment]
> This is exactly what I'm speaking about; when initializing, you can't know if the third dict refers to many edges and the keys are IDs or it refers to just one edge and the keys are attributes. 

Oh yeah currently the behaviour **is** broken, I am advocating for making it accept third dict as edge IDs by default and IMO the third layer should always be the edge_id because that is what the networkx MultiGraph data structure expects.

The following code should work after this behaviour is fixed.
``` python
In [14]: h = networkx.MultiGraph()

In [15]: h.add_edges_from([('a', 'b', {"w": 200}), ('a', 'b', {"w": 201})])
Out[15]: [0, 1]

In [16]: g = networkx.MultiGraph(networkx.to_dict_of_dicts(h))

In [17]: networkx.is_isomorphic(g, h)
Out[17]: True
```
and the example given in the original issue will be fixed by https://github.com/networkx/networkx/pull/4823
``` python
g = networkx.MultiGraph({'a': {'b': {0: {'w': 200}, 1: {'w': 201}}}})
print(g.edges)
h = networkx.MultiGraph()
h.add_edges_from([('a', 'b', {"w": 200}), ('a', 'b', {"w": 201})])
print(h.edges)

output

[('a', 'b', 0), ('a', 'b', 1)]
[('a', 'b', 0), ('a', 'b', 1)]
```
I am not sure why should we have `multigraph_flag` for every edge. One *can* add `0:{"attr":"foo"},1:{"attr":"bar"}` as an edge attribute but that would require people going in and changing the underlying `_adj` dictionary, but even after that the structure would be something like this
``` python
In [46]:  G = nx.MultiGraph()

In [47]: G.add_edge(1, 2)
Out[47]: 0

In [48]: G._adj[1][2][0] = {0:{"attr":"foo"}, 1:{"attr":"bar"}}

In [49]: G._adj
Out[49]:
{1: {2: {0: {0: {'attr': 'foo'}, 1: {'attr': 'bar'}}}},
 2: {1: {0: {0: {'attr': 'foo'}, 1: {'attr': 'bar'}}}}}
```

Something like the given example
```
nx.MultiGraph({'a':{'b':{'multigraph_flag':True,0:{"attr":"foo"},1:{"attr":"bar"}}},
'c':{'b':{'multigraph_flag':False (or not specifying) ,0:{"attr":"foo"},1:{"attr":"bar"}})
```
would be mixing MultiGraphs (d-o-d-o-d-o-d) with Graphs (d-o-d-o-d) and I am not sure if we should support that.

**Comment by dschult** [comment]
I think a d-o-d passed into the constructor should consist of only one type of data -- if you want to add edges that are different types, use `G.add_edges_from`...  It is already too complicated to figure out what `incoming_graph_data` is. :}

But there **are** cases where we want to input a d-o-d-o-d to a `nx.MultiGraph` constructor class. That can currently be done using `nx.to_networkx_graph(dodod, create_using=nx.MultiGraph)`, or `nx.MultiGraph(dodod)`.

The proposed "fix" in #4823 will hopefully allow for this same syntax to work as before...  BUT:  we also want to allow dododod inputs that have in the 3rd layer **either** edge_keys **or** not...  (This is slightly different from what @MridulS is promoting above.)  I think we agree that a d-o-d-o-d-o-d with edge_key in the 3rd layer should be accepted by the constructor and treated like the data structure of a MultiGraph.  Here is the other case where the d-o-d-o-d-o-d doesn't use an edge_key in the 3rd layer.

```python
dod = {'a': {'b': {'graphics': {'color': 'blue', 'shape': 'box'}, 'traits': {'budget': 300, 'revenue': 10}}}}

H1 = nx.MultiGraph(dod)
print(list(H1.edges))
H2 = nx.MultiGraph(dod, multigraph_input=False)
print(list(H2.edges))

output

[('a', 'b', 'graphics'), ('a', 'b', 'traits)]
[('a', 'b', 0)]
```
Note that for `H2`, the edge data has values that are themselves dicts... So the H2 data structure could be thoughts of as a d-o-d-o-d-o-d-o-d    (which is hard to type!)

**Comment by dschult** [comment]
I believe this issue is now fixed by #4823
